dev/bots/test.dart

```diff
@@ -176,6 +210,55 @@ Future<Null> _pubRunTest(
   return _runCommand(pub, args, workingDirectory: workingDirectory);
 }
 
+class EvalResult {
+  EvalResult({
+    this.stdout,
+    this.stderr,
+  });
+
+  final String stdout;
+  final String stderr;
+}
+
+Future<EvalResult> _evalCommand(String executable, List<String> arguments, {
+  String workingDirectory,
+  Map<String, String> environment,
+  bool skip: false,
+}) async {
+  final String commandDescription = '${path.relative(executable, from: workingDirectory)} ${arguments.join(' ')}';
+  final String relativeWorkingDir = path.relative(workingDirectory);
+  if (skip) {
+    _printProgress('SKIPPING', relativeWorkingDir, commandDescription);
+    return null;
+  }
+  _printProgress('RUNNING', relativeWorkingDir, commandDescription);
+
+  final Process process = await Process.start(executable, arguments,
+    workingDirectory: workingDirectory,
+    environment: environment,
+  );
+
+  final Future<List<List<int>>> savedStdout = process.stdout.toList();
+  final Future<List<List<int>>> savedStderr = process.stderr.toList();
+  final int exitCode = await process.exitCode;
+  final EvalResult result = new EvalResult(
+    stdout: UTF8.decode((await savedStdout).expand((List<int> ints) => ints).toList()),
+    stderr: UTF8.decode((await savedStderr).expand((List<int> ints) => ints).toList()),
+  );
+
+  if (exitCode != 0) {
```

Comment:
This seems to assume that the last 'RUNNING' message will inform the reader about the error.  I don't think you can assume that's true, since the processes run asynchronously. If that's correct, then it would be best to repeat the relativeWorkingDir, commandDescription (and the environment) on failure. It's probably worthwhile always printing this stuff anyway, since there might be tons of program output above the error message.

Reply:
Done.